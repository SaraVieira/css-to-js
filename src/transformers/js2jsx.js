import { formatProps, parseObj } from "../utils";

/**
 * Transforms a JS object to React props in JSX format.
 * @param {string} objString JS object code
 * @returns {string} transformed code
 */
export function transform(objString) {
  const rules = parseObj(objString);

  // Convert JS object to array of strings
  const propStrings = Object.keys(rules).map(key => {
    let value = rules[key];

    // The value `true` is special, because it can be transformed to a prop
    // with no value in JSX
    if (value === "true") {
      return key;
    }

    if (value.startsWith(`"`) || value.startsWith(`'`)) {
      // Value is probably a plain string, but need to do some more checks
      let first = value.charAt(0);
      let middle = value.slice(1, -1); // all except first and last char
      let last = value.charAt(value.length - 1);

      if (last !== first) {
        // It's a string but not correctly wrapped in quotes
        return `${key}={undefined}`;
      }

      if (middle.includes("\\") || middle.includes(`"`)) {
        // If the string value contains escaped characters, we need to wrap it in
        // curly brackets (escaping does not work in JSX prop values)
        value = `{${value}}`;
      } else {
        // Replace wrapping single quotes with double quotes
        if (first === `'`) {
          first = `"`;
        }
        if (last === `'`) {
          last = `"`;
        }
        value = `${first}${middle}${last}`;
      }
    } else {
      // Value is not a plain string, wrap it in curly braces
      value = `{${value}}`;
    }

    return `${key}=${value}`;
  });

  const props = propStrings.join(" ");
  try {
    return formatProps(props);
  } catch (e) {
    return props;
  }
}
