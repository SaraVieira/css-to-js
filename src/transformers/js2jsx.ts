import { formatProps } from "../utils/formatting";

/**
 * Transforms a JS object to React props in JSX format.
 * @param {string} objString JS object code
 * @returns {string} transformed code
 */
export function transform(objString) {
  // Loosly parse the code as a JS object
  const rules = {};
  objString.split("\n").forEach(line => {
    line = line.replace(/,$/, ""); // remove trailing comma

    // Split each line into a key and a value
    // Everything before the first : is the key and the reset is the value
    const segments = line.split(":");
    if (segments.length < 2) return; // skip this line
    const key = segments[0].trim();
    const value = segments
      .slice(1)
      .join(":")
      .trim();
    rules[key] = value;
  });

  // Convert JS object to array of strings
  const propStrings = Object.keys(rules).map(key => {
    let value = rules[key];

    // The value `true` is special, because it can be transformed to a prop
    // with no value in JSX
    if (value === "true") {
      return key;
    }

    if (value.startsWith(`"`) || value.startsWith(`'`)) {
      // Value is probably a plain string, but need to do some more checks
      let first = value.charAt(0);
      let middle = value.slice(1, -1); // all except first and last char
      let last = value.charAt(value.length - 1);

      if (last !== first) {
        // It's a string but not correctly wrapped in quotes
        return `${key}={undefined}`;
      }

      if (middle.includes("\\") || middle.includes(`"`)) {
        // If the string value contains escaped characters, we need to wrap it in
        // curly brackets (escaping does not work in JSX prop values)
        value = `{${value}}`;
      } else {
        // Replace wrapping single quotes with double quotes
        if (first === `'`) {
          first = `"`;
        }
        if (last === `'`) {
          last = `"`;
        }
        value = `${first}${middle}${last}`;
      }
    } else {
      // Value is not a plain string, wrap it in curly braces
      value = `{${value}}`;
    }

    return `${key}=${value}`;
  });

  try {
    return formatProps(propStrings.join(" "));
  } catch (e) {
    return "Could not generate valid props";
  }
}
