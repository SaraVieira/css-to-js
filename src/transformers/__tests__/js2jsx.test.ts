import { transform } from "../js2jsx";

describe("js2jsx", () => {
  test("transforms a rule with a string value", () => {
    const input = `someProp: "someValue"`;
    expect(transform(input)).toBe(`someProp="someValue"`);
  });

  test("transforms a rule with a string value with single quotes", () => {
    const input = `someProp: 'someValue'`;
    expect(transform(input)).toBe(`someProp="someValue"`);
  });

  test("transforms a rule with a number value", () => {
    const input = `someProp: 42`;
    expect(transform(input)).toBe(`someProp={42}`);
  });

  test("transforms a rule with an arbitrary expression value", () => {
    const input = `someProp: someExpression`;
    expect(transform(input)).toBe(`someProp={someExpression}`);
  });

  test("transforms a rule with value `true` to a prop with no value", () => {
    const input = `someProp: true`;
    expect(transform(input)).toBe(`someProp`);
  });

  test("transforms a simple object", () => {
    expect(
      transform(`{
        display: "block",
        fontSize: 16
      }`)
    ).toMatchSnapshot();
  });

  test("transforms a more complex object", () => {
    expect(
      transform(`{
      display: "block",
      margin: { sm: 4, md: 8 },
      padding: [2, 3],
      background: "#1e2f5d",
      fontSize: 16,
      fontFamily: "'Inter', sans-serif",
    }`)
    ).toMatchSnapshot();
  });

  test("transforms a rule with a double quote char in its value", () => {
    const input = `someProp: 'some"Value'`;
    expect(transform(input)).toBe(`someProp={'some"Value'}`);
  });

  test("transforms a rule with an escaped char in its value", () => {
    const input = `someProp: "some\\tValue"`;
    expect(transform(input)).toBe(`someProp={"some\\tValue"}`);
  });

  test("transforms an invalid string value to `{undefined}`", () => {
    const input = `someProp: "imInvalid`;
    expect(transform(input)).toBe(`someProp={undefined}`);
  });
});
