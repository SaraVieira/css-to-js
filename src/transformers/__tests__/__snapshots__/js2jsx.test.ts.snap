// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`js2jsx transforms a more complex object 1`] = `
"display=\\"block\\"
margin={{ sm: 4, md: 8 }}
padding={[2, 3]}
background=\\"#1e2f5d\\"
fontSize={16}
fontFamily=\\"'Inter', sans-serif\\""
`;

exports[`js2jsx transforms a simple object 1`] = `"display=\\"block\\" fontSize={16}"`;
